syntax = "proto3";

package kasir;

// Definisi request dan response untuk transaksi
message TransactionRequest {
    string item = 1;
    int32 jumlah = 2;
    float harga_per_unit = 3;
}

message TransactionResponse {
    float total_harga = 1;
    float diskon = 2;
    float pajak = 3;
    float harga_akhir = 4;
    string message = 5;
}

// Definisi request dan response untuk stok
message StockQuery {
    string item = 1;
}

message StockResponse {
    string item = 1;
    int32 stok = 2;
    string message = 3;
}

message StockUpdateRequest {
    string item = 1;
    int32 jumlah = 2;
}

message StockUpdateResponse {
    string message = 1;
}

// Definisi request dan response untuk laporan transaksi
message TransactionQuery {
    string start_date = 1;
    string end_date = 2;
}

message Transaction {
    int32 id = 1;
    string item = 2;
    int32 jumlah = 3;
    float harga_akhir = 4;
    string tanggal_transaksi = 5;
}

message TransactionReport {
    repeated Transaction transaksi = 1;
}

// Definisi layanan gRPC Backend
service Backend {
    rpc ProcessTransaction (TransactionRequest) returns (TransactionResponse);
    rpc CheckStock (StockQuery) returns (StockResponse);
    rpc UpdateStock (StockUpdateRequest) returns (StockUpdateResponse);
    rpc GetTransactionReport (TransactionQuery) returns (TransactionReport);  // RPC untuk laporan transaksi
}
